### PR TITLE
SCA dependency-graph safety hardening

### PR DESCRIPTION
**Pull request by endor-matt**
Hardens the SCA remediation agent's dependency-graph handling across all 13 supported package managers (8 ecosystem families: Maven, Gradle, npm, pip, Go, NuGet, Bundler, Cargo). Adds a per-manager mechanism table + Dependency Graph Safety Audit contract so manifest/lockfile manipulations are mechanism-driven and bounded per family — no cross-ecosystem guessing, no unverified removal/substitution claims — with per-family tests and fixtures backing each rule.

Validated by two full runtime-QA matrices (claude/sonnet + codex/gpt-5.6-luna, 8 ecosystem-balanced fixtures each): zero functional failures on either host across the hardened lanes.

### CODE REVIEW

**Review by dannykim-endor — COMMENTED** [review]
Verified before commenting:

- full suite: 945 passed, 0 failed
- `endor-agent-kit publish ... --prune --include-plugins` regenerates the tree byte-identical, so none of the ~200 generated files are hand-edited
- branch coverage over the SCA suites: `_base.py` 94%, every profile module 100%

Two before merge:

1. `_base.py:327` plus `_implementation.py:358` - the audited manifest is unconstrained when the selection omits its manifest list
2. `instructions.md:925` - the JSON skeleton still says `maven | gradle`

Rest is inline and non-blocking.

Two that have no line to anchor to:

- `docs/guardrails.md:280` "SCA gate checks include:" doesn't mention the dependency graph audit
- `CONTEXT.md` has no entry for `PackageManagerAuditProfile` or "dependency graph audit"

lmk if any of this doesn't match your read.

File: src/endor_agent_kit/workflow_output_contracts/sca/package_managers/_base.py [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+        )
+    audit_manifest = _text(audit.get("manifest"))
+    if (
+        audit_status != "unavailable"
+        and selected_manifests
+        and audit_manifest not in selected_manifests
+    ):
+        errors.append(
+            "dependency_graph_audit.manifest: must match a selected remediation manifest"
+        )
```
Comment (dannykim-endor) [inline src/endor_agent_kit/workflow_output_contracts/sca/package_managers/_base.py]: `audit.manifest` has no requiredness check anywhere, and `and selected_manifests` short-circuits the membership test, so an empty set lets any string through as the audited manifest.

```suggestion
    audit_manifest = _text(audit.get("manifest"))
    if audit_status != "unavailable":
        if not audit_manifest:
            errors.append(
                "dependency_graph_audit.manifest: required unless status is unavailable"
            )
        elif selected_manifests and audit_manifest not in selected_manifests:
            errors.append(
                "dependency_graph_audit.manifest: must match a selected remediation manifest"
            )
```

Keeps `selected_manifests` on the membership branch only, so the no-selection path at `_implementation.py:259` still works. Pairs with the change on `_implementation.py:358`.

File: src/endor_agent_kit/workflow_output_contracts/sca/_implementation.py [hunk 2]
```diff
@@ -178,15 +198,221 @@ def validate_sca_gate_payload(payload: dict[str, Any], *, gate: str = "selection
...
+                        f"{profile.display_name} remediations"
+                    )
+                    continue
+                selected_manifests = {
+                    item
+                    for source in (selected, selected_option)
+                    for field_name in ("manifests", "affected_manifests")
+                    for item in _list(source.get(field_name))
+                    if isinstance(item, str) and item.strip()
+                }
```
Comment (dannykim-endor) [inline src/endor_agent_kit/workflow_output_contracts/sca/_implementation.py]: This set is built only from `selected_remediation.manifests` / `affected_manifests`, and neither is required by any validator. `change_requests[0].inventory.key.manifest` is required at `:576` and isn't used here, so the set can be empty on an otherwise valid payload.

```suggestion
                selected_manifests = {
                    item
                    for source in (selected, selected_option)
                    for field_name in ("manifests", "affected_manifests")
                    for item in _list(source.get(field_name))
                    if isinstance(item, str) and item.strip()
                }
                for request in _list(payload.get("change_requests")):
                    if not isinstance(request, dict):
                        continue
                    inventory_manifest = _text(
                        _dict(_dict(request.get("inventory")).get("key")).get("manifest")
                    )
                    if inventory_manifest:
                        selected_manifests.add(inventory_manifest)
```

With both changes applied, an audit pointing at an unselected file and an audit with no manifest are both rejected; existing payloads are unaffected (304 SCA tests still pass).

File: source/agents/sca-remediation/instructions.md [hunk 3]
```diff
@@ -717,6 +921,14 @@ table, or other prose outside the object.
     "source_usage_summary": "required when CIA is indeterminate, risk is elevated, conflicts exist, or findings are introduced",
     "validation_requirements": []
   },
+  "dependency_graph_audit": {
+    "package_manager": "maven | gradle",
```
Comment (dannykim-endor) [inline source/agents/sca-remediation/instructions.md]: Stale since `0d882aac`. The prose enum at `:431` lists all 13; this skeleton is what prompt-only hosts copy, and it's propagated to the five generated artifacts (claude-code, codex, both gemini, portable).

```suggestion
    "package_manager": "maven | gradle | npm | yarn | pnpm | pip | poetry | pipenv | uv | go | nuget | bundler | cargo",
```

Needs a regen after. Worth a test asserting the skeleton's tokens equal `{p.name for p in SUPPORTED_PROFILES}` so it can't drift again.

File: src/endor_agent_kit/workflow_output_contracts/sca/_implementation.py [hunk 4]
```diff
@@ -178,15 +198,221 @@ def validate_sca_gate_payload(payload: dict[str, Any], *, gate: str = "selection
...
+        dependency_graph_audit = payload.get("dependency_graph_audit")
+        audit_present = isinstance(dependency_graph_audit, dict)
+        if dependency_graph_audit is not None and not audit_present:
+            # A dangerous audit wrapped in a list or emitted as a string
+            # must never ride through shape coercion.
+            errors.append("dependency_graph_audit: must be an object")
+        # A supplied audit is always validated, even with no selection: a
+        # dangerous audit must never ride through because the selection was
+        # nulled out.
+        if has_selected_remediation or audit_present:
```
Comment (dannykim-endor) [inline src/endor_agent_kit/workflow_output_contracts/sca/_implementation.py:256]: The audit requirement only fires inside the detections loop. Detection tests membership; `_validate_change_request_inventory` at `:576` tests presence. An `ecosystem`, `normalized_package`, and `manifest` that are all non-empty but match no profile satisfy the required-fields check and produce no detection, so no audit is demanded.

Needs a supported manager whose edited file isn't a conventional manifest name - a Maven `.properties` holding the version, a custom aggregator, a generated manifest.

The non-canonical token check at `:346` is also inside the loop, so a wrong ecosystem token on its own is never flagged.

Suggested hardening: reject an inventory whose `ecosystem` matches no profile alias.

File: src/endor_agent_kit/workflow_output_contracts/sca/package_managers/_base.py [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+                "mediation evidence"
+            )
+        if kind == "removal" and classification not in REMOVAL_CLASSIFICATIONS:
+            errors.append(
+                f"{prefix}.classification: {display} exclusions require unverified, "
+                "replacement_declared, replacement_verified, not_needed_verified, or "
+                "replacement_conflict_or_incomplete"
+            )
+        if kind == "substitution" and classification not in SUBSTITUTION_CLASSIFICATIONS:
+            errors.append(
```
Comment (dannykim-endor) [inline src/endor_agent_kit/workflow_output_contracts/sca/package_managers/_base.py:488]: Branch coverage puts `_base.py` at 94%, and the misses cluster on substitution:

- `:487` classification whitelist
- `:527` semantic_effect coupling
- `:557` validation_requirements
- `:579` `replacement_verified` -> `validated`

Each has a removal and an override sibling that is tested; the substitution twin isn't. Same at `:617` (verified removal replacement) and `:656` (validated override needing passed validation kinds).

`replacement_conflict_or_incomplete` has zero hits in `tests/` or `evals/cases.yaml` and drives two status-forcing rules at `:561` and `:583`.

I have tests covering these locally that take `_base.py` to 99% with zero missing statements - say the word and I'll attach them here or open a follow-up.

File: src/endor_agent_kit/structured_output_contracts.py [hunk 6]
```diff
@@ -904,6 +906,93 @@ def _risk_decision_schema() -> dict[str, Any]:
...
+                    "native_version_control",
+                    "forced_version_mediation",
+                    "dependency_removal",
+                    "dependency_substitution",
+                    "asset_or_feature_suppression",
+                    "source_override",
+                    "lockfile_override",
+                )
+            ),
+            "mechanism": _nullable_string(),
```
Comment (dannykim-endor) [inline src/endor_agent_kit/structured_output_contracts.py:944]: This block and `_base.py:21-88` are two sources of truth for the same vocabulary: `status`, `classification`, `semantic_effect`, `package_manager`, the Maven `type` enum, `["resolved_graph","runtime_linkage"]`, and the `maxItems` caps.

No test imports the `_base.py` constants. `tests/test_profile_contracts.py:89` asserts literal `12` and `8` against literal `12` and `8`; `:104` pins `semantic_effect` as a third literal copy; `status`, `classification`, and `package_manager` have no assertion at all.

Deriving the schema from the constants closes it.

Separately: `mechanism` here is `_nullable_string()` and `type` holds only Maven's 5 tokens, so the other 12 managers are enforced by the runtime validator alone. Confirming that's deliberate.

File: src/endor_agent_kit/workflow_output_contracts/sca/_implementation.py [hunk 7]
```diff
@@ -178,15 +198,221 @@ def validate_sca_gate_payload(payload: dict[str, Any], *, gate: str = "selection
...
+                selected_manifests = {
+                    item
+                    for source in (selected, selected_option)
+                    for field_name in ("manifests", "affected_manifests")
+                    for item in _list(source.get(field_name))
+                    if isinstance(item, str) and item.strip()
+                }
+                selected_package_names = set()
+                selected_vulnerable_versions = set()
+                for source in (selected, _dict(payload.get("selected_option"))):
```
Comment (dannykim-endor) [inline src/endor_agent_kit/workflow_output_contracts/sca/_implementation.py]: `selected_manifests`, `selected_package_names`, and `selected_vulnerable_versions` are rebuilt inside `for detection in detections:` but don't depend on `detection` - they can hoist above the loop. This line also re-derives `selected_option`, which is already local at `:201`.

Those three sets plus `risk_status` travel together into `validate_dependency_graph_audit` as four keyword args.

File: src/endor_agent_kit/workflow_output_contracts/sca/_implementation.py [hunk 8]
```diff
@@ -198,6 +424,51 @@ def validate_sca_gate_payload(payload: dict[str, Any], *, gate: str = "selection
     return errors
 
 
+def _detect_selected_package_managers(
+    payload: dict[str, Any],
```
Comment (dannykim-endor) [inline src/endor_agent_kit/workflow_output_contracts/sca/_implementation.py:445]: Two here:

1. no return type on this function, while its neighbors have one. No ruff or mypy config in the repo, so nothing catches it.
2. `:26-29` imports `_fold_disguises` and `_normalize_version_token` from `package_managers._base`, bypassing the `__all__` at `__init__.py:61-84`. Practice elsewhere imports private modules, not private functions across a package boundary. Promoting those two onto the package's public surface keeps the seam intact.

File: src/endor_agent_kit/workflow_output_contracts/sca/_implementation.py (hunk 4 above)
Comment (endor-matt) [reply src/endor_agent_kit/workflow_output_contracts/sca/_implementation.py:256]: Took the fail-close, different shape: rejecting ecosystems with no profile alias would block managers we support but haven't given an audit profile (composer). Instead, zero-detection selections now require the audit object — the unavailable-and-empty shape passes, anything else fails closed. Tests cover both sides. resolved in latest commit

File: src/endor_agent_kit/structured_output_contracts.py (hunk 6 above)
Comment (endor-matt) [reply src/endor_agent_kit/structured_output_contracts.py:944]: Schema now derives from the _base.py constants (lazy import keeps the module a leaf; frozensets sorted for deterministic JSON), plus a lockstep test. And yes — nullable mechanism / Maven-only type is deliberate: strict schemas can't do per-manager conditionals, so that's runtime-validator-owned. Comment added at the site.

Comment (endor-matt) [reply src/endor_agent_kit/structured_output_contracts.py:944]: resolved in latest commit

**Comment by endor-matt** [comment]
@dannykim-endor resolved everything in this latest commit, good catches

**Review by dannykim-endor — APPROVED** [review]
Approving. Re-verified all eight items against the new commit, and independently re-ran the ones that mattered rather than reading them.

Confirmed fixed: manifest requiredness and the inventory-manifest anchor (probed both the laundering and missing-manifest shapes), the skeleton enum plus the parity test, the zero-detection fail-open, the substitution and `replacement_conflict_or_incomplete` coverage, and the schema derivation. `_base.py` is at 100% statement and branch coverage, 963 tests pass, no import cycle from the lazy import, and the derived schema is byte-identical across hash seeds.

On the zero-detection rule - you were right to reject my "reject unknown ecosystem alias" suggestion. I checked, and managers we support without an audit profile would have been blocked by it. Your shape is better, and the `raw_manipulations not in (None, [])` guard against `_list` coercion is a good catch.

Two things I'm tracking separately rather than holding the PR for, both non-blocking:

1. Closing the omit-the-audit path means unprofiled managers must now report `unavailable`, which trips the pre-existing `unavailable` / `approved_low_risk` coupling. Net effect is that Composer and Swift remediations can't reach `approved_low_risk` at any gate, so the PR body always renders the neutral expectation line. Defensible as the conservative answer, but it's an emergent consequence rather than a decision anyone made - worth settling deliberately.

2. `test_unsupported_ecosystem_unavailable_audit_passes` pins `"package_manager": "conan"`, but the transport enum is the 13 profile names plus null, so a schema-constrained host would reject that exact object. `null` passes both the runtime validator and the schema. Worth switching the test and naming the expected value in the instructions.

Neither needs to land here. Nice work on the turnaround.
